Add Search endpoint

pmn4 opened the pull request:
This PR adds the "Search" endpoint. Search doesn't quite fit the Resourceful pattern, so I put the tests in `client_spec.rb`. Happy to move them if you'd prefer them to be in their own file (I just felt strange because there is no resource to use in `describe <Resource>`).

I also added a new Error class, and to allow for better rescuing, created an error superclass for the gem. Also happy to revert if it doesn't feel right.

pmn4 commented on lib/iex/errors/symbol_not_found_error.rb:
```diff
@@ -1,14 +1,27 @@
   module Errors
-    class SymbolNotFoundError < StandardError
-      attr_reader :symbol
+    class IexError < StandardError
```
added an Error superclass to allow for other projects to:
``` Ruby
rescue IEX::Errors::IexError
```
instead of having to:
``` Ruby
rescue IEX::Errors::SymbolNotFoundError, IEX::Errors::SearchNotFoundError, ... # as more errors are added
```

pmn4 replied:
tbh, I hadn't realized there were other errors defined (I thought I was creating the second one). If I had, I probably would have put this into a separate PR. sorry for the confusion!

pmn4 commented on spec/iex/client_spec.rb:
```diff
@@ -135,4 +135,36 @@
     end
   end
+
+  context '#search' do
```
tests for this endpoint were added to `client_spect.rb` because other endpoints have associated resources, where `#search` (eventually, according to the documentation) returns any type of resource.

dblock replied:
Make a directory, `search` and create `search/quote_spec.rb` and such? For the general use case of errors make a `search_spec.rb`.

dblock commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 * Your contribution here.
 * [#57](https://github.com/dblock/iex-ruby-client/pull/57): Update properties for chart api endpoint - [@brunjo](https://github.com/brunjo).
+* [#58](https://github.com/dblock/iex-ruby-client/pull/58): Add search endpoint - [@pmn4](https://github.com/pmn4)
```
Missing period at the end to stop danger from complaining.

pmn4 replied:
🤣  on it!

dblock commented on lib/iex/endpoints/search.rb:
```diff
@@ -0,0 +1,20 @@
+  module Endpoints
+    module Search
+      def search(fragment, options = {})
+        # consider raising if fragment.blank?
```
You should, raise `ArgumentError`.

dblock commented on lib/iex/errors/symbol_not_found_error.rb:
```diff
@@ -1,14 +1,27 @@
   module Errors
-    class SymbolNotFoundError < StandardError
-      attr_reader :symbol
+    class IexError < StandardError
```
Should this be IEXError? Or just Error since it's already in the namespace?

pmn4 replied:
I like just `Error`, will update.

dblock commented on spec/fixtures/iex/client/search/missing.yml:
```diff
@@ -0,0 +1,49 @@
+http_interactions:
+- request:
+    method: get
+    uri: https://sandbox.iexapis.com/v1/search/?token=Tsk_341c973d296e4e18aa61dd63050ce235
```
Remove your actual token from here.

pmn4 replied:
SO DUMB! thanks. I regenerated my tokens just to be safe

dblock commented on spec/iex/client_spec.rb:
```diff
@@ -135,4 +135,36 @@
+      # order for VCR to record the response, we need the endpoint to work
+      client.endpoint = 'https://sandbox.iexapis.com/v1'
+      # using the documentation-provided token
+      client.publishable_token = 'Tsk_341c973d296e4e18aa61dd63050ce235'
```
We default this to `token` and replace it everywhere for tests, as they are run from a recording. Don't leak your tokens!

pmn4 replied:
This particular endpoint is not available to free tier accounts (the plan I'm on!), so I had to update the endpoint to the sandbox, where of course, different keys are needed. the token in this code is copy/pasted from the iexcloud docs, so it's already publicly exposed!

dblock replied:
Is it that they have it in sandbox but not in production yet? In which case you can manually remove the sandbox part from tests and edit VCRs once you have a working recording.

dblock commented on spec/iex/client_spec.rb:
```diff
@@ -135,4 +135,36 @@
+    it 'handles missing search terms', vcr: { cassette_name: 'client/search/missing' } do
+      expect { client.search('') }.to raise_error \
+        IEX::Errors::SearchNotFoundError,
+        'Fragment  Not Found'
```
The fact that there's a double space in the expected text looks suspicious.

dblock commented on lib/iex/endpoints/search.rb:
```diff
@@ -0,0 +1,20 @@
+          'search',
+          fragment
+        ].compact.join('/'), options).map do |data|
+          # Per the IEX documentation, any type of Resource could be returned
```
Yes, please. And create a base resource for anything unsupported.

pmn4 replied:
fyi, they have not released this feature yet, the docs just mention that it is coming

dblock replied:
I don't see a problem supporting something in pre-release, but we have to say it in README and point to whatever IEX docs that say the same

dblock commented:
The code here looks good. You still need to remove your actual secret token from everywhere!

dblock commented:
Also add a comment when this is ready to re-review, commits don't get flagged in notifications.

dblock commented:
Bump @pmn4, want to finish this?

pmn4 commented:
@dblock there are a few items here:

1) Resource Type: I introduced some confusion with the search results "Resource Type." IEX has not yet launched this feature, so there is nothing to support, or even details about _how_ to support it. The documentation says it's coming, so I left a comment mentioning that it is coming.

2) Tokens: the token in this PR is copy/pasted from IEX's documentation (it is their sandbox API key). I used it because the search endpoint is a paid feature, for which I do not have an API key. Instead, for this one test case, I updated both the endpoint and API key to point to the sandbox environment, so if we need to recreate the VCR response, we can without having to pay.

do these make sense? sorry for the confusion, and the delay! (tbh, I was only evaluating IEX when I authored this PR, and have since switched gears to another provider.)

dblock commented:
> @dblock there are a few items here:
> 
> 1. Resource Type: I introduced some confusion with the search results "Resource Type." IEX has not yet launched this feature, so there is nothing to support, or even details about _how_ to support it. The documentation says it's coming, so I left a comment mentioning that it is coming.

ok

> 2. Tokens: the token in this PR is copy/pasted from IEX's documentation (it is their sandbox API key). I used it because the search endpoint is a paid feature, for which I do not have an API key. Instead, for this one test case, I updated both the endpoint and API key to point to the sandbox environment, so if we need to recreate the VCR response, we can without having to pay.

you just don't want the real token in this PR because someone will steal it; once a recording was made you can just replace it with a placeholder

dblock commented:
> do these make sense? sorry for the confusion, and the delay! (tbh, I was only evaluating IEX when I authored this PR, and have since switched gears to another provider.)

I'll leave this open for someone/you to finish it.
